Support setting user Codespaces secrets

joshmgross opened the pull request:
Supports #4698


This PR adds support for setting a Codespaces user secret via the CLI.

```bash
gh secret set -u -bs -vselected -r"cli/cli,github/hub" cool_secret
```

Codespaces user secrets follow a similar pattern to Actions organization secrets in that they can be shared with all repositories or a specific set of repositories.

To support this, this PR uses the public key and set secret APIs for user Codespaces secrets.

https://docs.github.com/en/rest/reference/codespaces#get-public-key-for-the-authenticated-user
https://docs.github.com/en/rest/reference/codespaces#create-or-update-a-secret-for-the-authenticated-user

### Repository selection

Unlike organization secrets, the repositories specified for a user secret do not have to belong to that owner of the secret.

When setting an organization secret via the CLI, the `--repos` option is used to specify a set of repository names.

For user secrets, I've reused that same `--repos` option but with the assumption that all repositories specified are fully qualified names with owners (NWOs). 

If this is confusing, we could either:
1. Use a different option for fully qualified repos rather than reusing `--repos`
2. Support both repository names and full NWOs. If an owner isn't specified, assume the owner is the current user.

I did not use the second option as I'm unsure how to retrieve the current user's login.

### Auth Issues

These APIs require the `user` or `read:user` scope, which is not currently requested by default.

Users will need to request the `user` scope to set a user secret
```bash
gh auth refresh -h github.com -s user
```

Since setting a secret first fetches the public key which requires `read:user` scopes and the set secret API requiring `user` scopes, users may hit auth issues twice before they are successful.

```bash

>gh secret set -u -bs -vselected -r"cli/cli,github/hub" cool_secret
failed to fetch public key: HTTP 403: This API requires the `read:user` scope. (https://api.github.com/user/codespaces/secrets/public-key)
This API operation needs the "read:user" scope. To request it, run:  gh auth refresh -h github.com -s read:user

>gh auth refresh -h github.com -s read:user 
...

>gh secret set -u -bs -vselected -r"cli/cli,github/hub" cool_secret
failed to set secret: HTTP 403: This API requires the `user` scope. (https://api.github.com/user/codespaces/secrets/cool_secret)
This API operation needs the "user" scope. To request it, run:  gh auth refresh -h github.com -s user

>gh auth refresh -h github.com -s user
...
```

### Testing

Validated that a secret was able to be added:
```bash
❯ bin/gh secret set -u -bs -vselected -r"joshmgross/clock,joshmgross/actions-testing" coolSecret
✓ Set secret coolSecret for cli/cli
```

![image](https://user-images.githubusercontent.com/16631042/141234644-e60f3278-65e8-4ac2-a2f9-7c38610bf906.png)

Fixes https://github.com/cli/cli/issues/4038

joshmgross committed:
Support setting user Codespaces secrets

mislav commented on pkg/cmd/secret/set/http.go:
```diff
@@ -14,10 +15,10 @@ import (
-	KeyID          string `json:"key_id"`
+	EncryptedValue string   `json:"encrypted_value"`
+	Visibility     string   `json:"visibility,omitempty"`
+	Repositories   []string `json:"selected_repository_ids,omitempty"`
```
If the endpoint for setting Codespaces secrets accepts a different type for `selected_repository_ids` than other endpoints, then I would suggest defining a separate SecretPayload type to be used for Codespaces secrets only. Right now, the change that you've made here isn't backwards-compatible.

joshmgross replied:
Fixed in https://github.com/cli/cli/pull/4699/commits/7d0e2bc771c8de4b2fc523d5f9e32404f0ff787b

mislav commented on pkg/cmd/secret/set/http.go:
```diff
@@ -75,10 +85,14 @@ func putOrgSecret(client *api.Client, host string, pk *PubKey, opts SetOptions,
-		repositoryIDs, err = mapRepoNameToID(client, host, orgName, opts.RepositoryNames)
+		repos := make([]ghrepo.Interface, 0, len(opts.RepositoryNames))
+		for _, repo := range opts.RepositoryNames {
+			repos = append(repos, ghrepo.New(opts.OrgName, repo))
```
Please use `ghrepo.NewWithHost` instead of `New`, which is soft-deprecated

joshmgross replied:
Changed as part of https://github.com/cli/cli/pull/4699/commits/e0d9b32d28b0bcbee91cc89085c7586a2db45f99

mislav commented on pkg/cmd/secret/set/http.go:
```diff
@@ -75,10 +85,14 @@ func putOrgSecret(client *api.Client, host string, pk *PubKey, opts SetOptions,
-		repositoryIDs, err = mapRepoNameToID(client, host, orgName, opts.RepositoryNames)
+		repos := make([]ghrepo.Interface, 0, len(opts.RepositoryNames))
+		for _, repo := range opts.RepositoryNames {
+			repos = append(repos, ghrepo.New(opts.OrgName, repo))
```
Since we now accept fully-qualified repo names when setting Codespaces secrets, let's also (optionally) accept fully qualified names when setting org secrets too. Of course, accepting just repo names should be kept for backwards compatibility.

```
gh secret set -o acme -r project1,project2
gh secret set -o acme -r acme/project1,acme/project2
```

joshmgross replied:
Added with https://github.com/cli/cli/pull/4699/commits/e0d9b32d28b0bcbee91cc89085c7586a2db45f99

mislav commented on pkg/cmd/secret/set/http.go:
```diff
@@ -114,14 +161,14 @@ func putRepoSecret(client *api.Client, pk *PubKey, repo ghrepo.Interface, secret
-func mapRepoNameToID(client *api.Client, host, orgName string, repositoryNames []string) ([]int, error) {
-	queries := make([]string, 0, len(repositoryNames))
-	for i, repoName := range repositoryNames {
+func mapRepoToID(client *api.Client, host string, repositories []ghrepo.Interface) ([]string, error) {
```
Nice. Love how this method is now generalized

mislav commented on pkg/cmd/secret/set/set.go:
```diff
@@ -121,6 +122,7 @@ func NewCmdSet(f *cmdutil.Factory, runF func(*SetOptions) error) *cobra.Command
 	}
 	cmd.Flags().StringVarP(&opts.OrgName, "org", "o", "", "Set a secret for an organization")
 	cmd.Flags().StringVarP(&opts.EnvName, "env", "e", "", "Set a secret for an environment")
+	cmd.Flags().BoolVarP(&opts.UserSecrets, "user", "u", false, "Set a secret for the current user")
```
Are these secrets used for anything else other than Codespaces? Can we call them "codespace secrets" instead of "user secrets"? To me, "user secret" sounds vague and doesn't evoke connection to Codespaces at all.

joshmgross replied:
Secrets are attached to an entity (user, org, repo, environment) and associated with some feature area/product (Actions, Codespaces, Dependabot).

Both Codespaces and Dependabot support secrets at the repository and organization level - 
https://docs.github.com/en/codespaces/managing-codespaces-for-your-organization/managing-encrypted-secrets-for-your-repository-and-organization-for-codespaces
https://docs.github.com/en/code-security/supply-chain-security/keeping-your-dependencies-updated-automatically/managing-encrypted-secrets-for-dependabot


Currently, the only non-Actions secrets with a API support are Codespaces user secrets.

What do you think about adding a new parameter to choose between Actions/Codespaces?
i.e.
`bin/gh secret set -u -codespaces`

or 

`bin/gh secret set -u --app=Codespaces`

We could add some validation to prevent someone from specifying Codespaces or Actions for an unsupported entity type

mislav replied:
Let's hold off on the new parameter and keep just `--user` for the sake of simplicity. Our added docs should clarify what "user secrets" are useful for at this moment. 👍

mislav reviewed (changes requested):
Thanks! This feature looks great.

Left some comments, but the biggest blocker remains the uncertainty about the type of the `selected_repository_ids` parameter. Let us know what you learn!

joshmgross commented:
Thanks for the review @mislav, I've filed issues with some internal teams to get the `selected_repository_ids` type issues fixed.

The Actions API is incorrectly documented and does expect those to be integers.

joshmgross committed:
Use a separate payload for Codespaces secrets

joshmgross committed:
Remove unused struct

joshmgross committed:
Support fully-qualified repo names for org secrets

joshmgross committed:
Merge branch 'trunk' into joshmgross/codespaces-set-secret

joshmgross commented on pkg/cmd/secret/secret.go:
```diff
@@ -15,11 +15,9 @@ func NewCmdSecret(f *cmdutil.Factory) *cobra.Command {
 			Secrets can be set at the repository, environment, or organization level for use in
-			GitHub Actions. Run "gh help secret set" to learn how to get started.
-		`),
-		Annotations: map[string]string{
```
Resolved conflicts with #4675, with this change secrets would no longer just be an Actions command. Let me know if there's a better tag here.

mislav committed:
secret set: do not look up base repo in case of user secret

mislav committed:
Do not accept nor submit `visibility` for user secrets

Visibility is only meant for org secrets. User secrets do support
scoping to specific repositories, but for that the `--repos` flag can be
used without passing `--visibility`.

mislav committed:
Remove client-side validation of secret names

Let the platform perform validation. We gain nothing from client-side
validation other than avoding an API request. By performing client-side
validation, we open ourselves to the risk of falling out of sync with
what the plaform considers valid secret names.

mislav committed:
Improve `secret set` docs to clarify different levels

mislav committed:
Improve flag parsing for `secret set`

Do not use `Flags().Changed(...)` when testing for whether a flag has a
non-empty value or not. The user might have intentionally sent an empty
value.

mislav committed:
Improves `gh secret set --repos` support for GitHub Enterprise

Ensure the `--repos` list is parsed with the host of the repository as
the default host in the case repositories are listed in the `repo` or
`owner/repo` formats (which are also most likely).

mislav reviewed (approved):
This looks great!

I have pushed fixes:
- `gh secret set --user` is now usable when _not_ in a context of a local git repository
- `--visiblity` is now unsupported for `--user` secrets
- do not send the `visibility` JSON parameter when setting user secrets since that's ignored by the API endpoint
- improve docs to clarify different secret levels (repo/environment/org/user)
- improve flag parsing and error formatting in `secret set`
- fix `secret set --repos` for GitHub Enterprise
- remove client-side validation of secret names (just for good measure)

@joshmgross The remaining items are now that `gh secret list` and `gh secret remove` do not support `--user`. Do you have time to add those too? I suspect it will be simple. Should that be added in a separate PR? You can base it on this branch.

joshmgross commented:
Thanks @mislav, your changes look good to me. I completely missed that user secrets didn't have visibility options like organization secrets.

I can add support for `gh secret list` and `gh secret remove` in a separate PR, but I don't think I'll be able to use this PR as a base branch since this is a fork pull request.

mislav commented:
@joshmgross Ah, right! The `list` and `remove` commands are completely separate Go packages, so you might be able to base your changes on `trunk` just as well, even before this PR gets merged.

joshmgross commented:
Adding `list` and `remove` in #4714
